src/Microsoft.DotNet.Cli.Utils/ProcessReaper.cs

```diff
@@ -0,0 +1,163 @@
+    /// process to handle the signal.
+    /// </remarks>
+    internal class ProcessReaper : IDisposable
+    {
+        /// <summary>
+        /// Creates a new process reaper.
+        /// </summary>
+        /// <param name="process">The target process to reap if the current process terminates. The process must already be started.</param>
+        public ProcessReaper(Process process)
+        {
+            _process = process;
+
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
+            {
+                _job = AssignProcessToJobObject(_process.Handle);
+            }
+            else
+            {
+                AppDomain.CurrentDomain.ProcessExit += HandleProcessExit;
+            }
+
+            Console.CancelKeyPress += HandleCancelKeyPress;
+        }
+
+        public void Dispose()
+        {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
+            {
+                if (_job != null)
+                {
+                    // Clear the kill on close flag because the child process terminated successfully
+                    // If this fails, then we have no choice but to terminate any remaining processes in the job
+                    SetKillOnJobClose(_job.DangerousGetHandle(), false);
+                    
+                    _job.Dispose();
+                    _job = null;
+                }
+            }
+            else
+            {
+                AppDomain.CurrentDomain.ProcessExit -= HandleProcessExit;
+            }
+
+            Console.CancelKeyPress -= HandleCancelKeyPress;
+        }
+
+        private static void HandleCancelKeyPress(object sender, ConsoleCancelEventArgs e)
+        {
+            // Ignore SIGINT/SIGQUIT so that the process can handle the signal
+            e.Cancel = true;
+        }
+
+        private static SafeWaitHandle AssignProcessToJobObject(IntPtr process)
+        {
+            var job = NativeMethods.Windows.CreateJobObjectW(IntPtr.Zero, null);
+            if (job == null || job.IsInvalid)
+            {
+                return null;
+            }
+
+            if (!SetKillOnJobClose(job.DangerousGetHandle(), true))
+            {
+                job.Dispose();
+                return null;
+            }
+
+            if (!NativeMethods.Windows.AssignProcessToJobObject(job.DangerousGetHandle(), process))
+            {
+                job.Dispose();
+                return null;
+            }
+
+            return job;
+        }
+
+        private void HandleProcessExit(object sender, EventArgs args)
+        {
+            var currentPid = Process.GetCurrentProcess().Id;
+            bool sendChildSIGTERM = true;
+
+            // First, try to SIGTERM our process group
+            // If the pgid is not the same as pid, then this process is not the root of the group
+            if (NativeMethods.Posix.getpgid(currentPid) == currentPid)
```

Comment:
There is no code that calls `setpgid. Why deal with this specifically?

Reply:
If this process is a child of another that doesn't `setpgid` (e.g. another .NET process that spawns `dotnet`), then this will be false.  It was to avoid a failing call to `kill` for the group.  If you think that's unnecessary and I should just call `kill` and check for `ESRCH` for the group, I'm fine with changing it.

Reply:
I'm wondering when this would be true, so my comment was more suggesting to remove the whole if block.

Reply:
Most shells `setpgid` though, right?

Reply:
Ah, right. If I start dotnet from a bash script though, they are not the same.
This may cause subtle differences in behaviour.
If we remove this, things will still work, and should be consistent independent of being process group leader or not.

Reply:
Agreed that there will be a difference in behavior when pgid != pid for the parent dotnet process (one SIGTERMs the process group and one SIGTERMs the immediate child only).

I'm also fine with removing the process group `kill` entirely and just consistently SIGTERMing the immediate child for complete consistency on POSIX systems.  I was just hoping to make it behave like it does on Windows in terms of terminating the child process tree by default.

I mean, it is trivial to signal a group when running `kill` from a shell manually, so I have no strong opinions either way.

If we do keep the process group `kill`, then I can remove the `getpgid` call in favor of just checking for `ESRCH` from the `kill` if we feel that's better.

Reply:
My preference is to either remove the pg kill, or to also do a setpgid.

Reply:
I'd be in favor of `setpgid` the child and just SIGTERM'ing it as a group consistently (assuming `setpgid` succeeds, otherwise we'll just SIGTERM the child pid).  I'll put up a PR with that change and also for the wait(0) on the child process object to guard against pid recycling.

Thanks again for the feedback!

Reply:
> My preference is to either remove the pg kill, or to also do a setpgid.

The first is the safest, and less like Windows.
The second is more like Windows, but may suffer from side-effects. I think SIGINT may no longer flow to the child.

Reply:
Oh good point.  I'd have to forward SIGINT in that case too.  Let me experiment.

Reply:
So unfortunately `setpgid` must occur before the `exec*` call (for valid reasons), and I'm not going to duplicate the complex process creation process that the framework does just for this fix.

As such, I think I'll remove the pg kill.

Reply:
> So unfortunately setpgid must occur before the exec* call (for valid reasons), and I'm not going to duplicate the complex process creation code

Setting it before calling `Process.Start` would be enough.
That said, I think we should keep it simple and not change the process group unless we need to.

Reply:
Doesn't `setpgid` require a `fork` prior to being called?  I thought the framework only forks as a result of `Process.Start` (seems to happen in `StartCore` in the Unix implementation.

Reply:
> Or did you meant to setpgid the parent dotnet process before spawning the child? 

That is what I meant, so the child would inherit it from the parent.

>  In the case where the dotnet process is not in its own group, then it would probably be unwise to break out.

> The only consistent way to do it would to be setpgid the child prior to the exec, which we can't do unless CoreFX implements dotnet/corefx#17412.

It's more tricky than I thought. If the dotnet process would change its own pgid, it would no longer receive signals (e.g. SIGINT) from the console. So setpgid needs to occur between fork and exec of the child. And the parent needs to take care of propagating the signals to the children.

Reply:
Plus, no managed code can safely run between fork and exec. If we ever need something there, we’d need API changes in the BCL.